sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs

```diff
@@ -184,19 +184,21 @@ private ProgramState ProcessStreamDisposingTypes(ProgramState programState, Synt
 
                 foreach (var objectCreationExpression in objectCreationExpressions)
                 {
-                    var objectType = this.semanticModel.GetSymbolInfo(objectCreationExpression.Type)
-                    .Symbol
-                    .GetSymbolType();
-
-                    if (objectType.DerivesOrImplementsAny(typesDisposingUnderlyingStream) &&
-                        objectCreationExpression.ArgumentList?.Arguments is IEnumerable<ArgumentSyntax> argumentList)
+                    if (this.semanticModel.GetSymbolInfo(objectCreationExpression.Type).Symbol is ISymbol symbol &&
+                        symbol.GetSymbolType() is ITypeSymbol objectType &&
+                        objectType.DerivesOrImplementsAny(typesDisposingUnderlyingStreamWithLeaveOpenArgumentIndex.Keys.ToImmutableArray()) &&
+                        objectCreationExpression.ArgumentList?.Arguments is IEnumerable<ArgumentSyntax> argumentList &&
+                        argumentList.Any())
                     {
                         // Checks for the 'leaveOpen' argument. It is at position 4 if present for StreamReader constructor, while it is at position 3 for StreamWriter.
                         // See #2491 for more details
-                        var leaveOpenArgumentIndex = objectType.DerivesOrImplements(KnownType.System_IO_StreamReader) ? 4 : 3;
-                        if (!HasLeaveOpenArgument(argumentList, leaveOpenArgumentIndex))
+                        var leaveOpenArgumentIndex = typesDisposingUnderlyingStreamWithLeaveOpenArgumentIndex

```

Comment:
isn't his a Dictionary, can't you just get the element by key?

Reply:
Retrieving the element by key is not the same as getting it for types that `type.DerivesOrImplements` one of these types. I did not want to change the logic, but we can be more strict and only consider these 2 exact types here

Reply:
sorry, I missed that